### PR TITLE
Storage implementation

### PR DESCRIPTION
**Pull request by EMattfolk**
I have started adding some of the hooks we will be using.
Comments are appreciated!

Also, we will be writing in english in these pull requests.

And yes, there are no comments yet.

### CODE REVIEW

**Comment by malin-mallan** [comment]
I get an error when trying you code.

image

**Comment by EMattfolk** [comment]
> I get an error when trying you code.

Did you `npm install` before running?

**Comment by malin-mallan** [comment]
> > I get an error when trying you code.
> 
> Did you `npm install` before running?

Oh I missed that... 
Now the code is working on both mobile and web!

**Comment by EMattfolk** [comment]
Plonk this in App.tsx to test (only works in web):

```javascript
import React from 'react';
import StorageTest from './src/storage/components/storageTest';

export default function App() {
  return (
      <StorageTest />
  );
}
```

**Comment by malin-mallan** [comment]
the test program works really good in the web!

File: src/storage/activities.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+   * Returns a copy of the activities object with specified date inserted.
+   *
+   * @remarks
+   * A date will not be inserted if already present.
+   *
+   * @param date - The date to be inserted
+   * @returns The activities object with specified date inserted
+   */
+  const _insertDay = (date: string): Activities => {
+    const newActivities = JSON.parse(JSON.stringify(activities));
```
Comment (Dolvur) [inline src/storage/activities.tsx:63]:
```suggestion
    const newActivities: Activities = JSON.parse(JSON.stringify(activities));
```

File: src/storage/activities.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+   *
+   * @param date - The date on which the entry will be inserted
+   * @param hour - The hour on which the entry will be inserted
+   * @param entry - The entry to be inserted
+   * @returns `true` if the entry was inserted, `false` otherwise
+   */
+  const add = (date: string, hour: number, entry: ActivitiesEntry): boolean => {
+    if (0 <= hour && hour < 24) {
+      const newActivities = _insertDay(date);
+      const index = newActivities.findIndex(a => a.date === date);
```
Comment (Dolvur) [inline src/storage/activities.tsx:91]:
```suggestion
      const newActivities: Activities = _insertDay(date);
      const index: number = newActivities.findIndex(activity => activity.date === date);
```

File: src/storage/activities.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+   *
+   * @remarks
+   * A date will not be inserted if already present.
+   *
+   * @param date - The date to be inserted
+   * @returns The activities object with specified date inserted
+   */
+  const _insertDay = (date: string): Activities => {
+    const newActivities = JSON.parse(JSON.stringify(activities));
+    if (!activities.some(a => a.date === date)) {
```
Comment (Dolvur) [inline src/storage/activities.tsx:64]:
```suggestion
    if (!activities.some(activity => activity.date === date)) {
```

File: src/storage/activities.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      const index = newActivities.findIndex(a => a.date === date);
+      newActivities[index].entries[hour] = entry;
+      AsyncStorage.setItem(activitiesKey, JSON.stringify(newActivities))
+        .then(() => setActivities(newActivities));
+      return true;
+    }
+    return false;
+  }
+
+  const modifyActivities = {
```
Comment (Dolvur) [inline src/storage/activities.tsx]:
```suggestion
  const modifyActivities: ModifyActivities = {
```

File: src/storage/activities.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  /**
+   * Inserts an entry into the activities object and updates AsyncStorage.
+   *
+   * @remarks
+   * This function will add the date if not present.
+   *
+   * @param date - The date on which the entry will be inserted
+   * @param hour - The hour on which the entry will be inserted
```
Comment (Dolvur) [inline src/storage/activities.tsx]:
```suggestion
   * @param hour - The hour on which the entry will be inserted (0 up to 23)
```

File: src/storage/activities.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+  const modifyActivities = {
+    add: add,
+  };
+
+  useEffect(() => {
+    AsyncStorage.getItem(activitiesKey)
+    .then(v => v === null ? activitiesDefault : JSON.parse(v))
+    .then(v => setActivities(v));
+  }, []);
```
Comment (Dolvur) [inline src/storage/activities.tsx]: If there is time, perhaps you could add a small description to this and the similar functions. Perhaps it should be higher up after useState also, so one can easily see that the state (activities in this case) might have been changed during the initial render when following the code? I'm not sure it's up to you.

File: src/storage/calendar.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+export const calendarDefault: Calendar = [];
+
+/**
+ * Compares two calendar entries and returns `true` if they are equal.
+ *
+ * @param a - The first calendar entry
+ * @param b - The second calendar entry
+ * @returns a === b
+ */
+const entryEq = (a: CalendarEntry, b: CalendarEntry) => {
```
Comment (Dolvur) [inline src/storage/calendar.tsx]:
```suggestion
const entryEq = (a: CalendarEntry, b: CalendarEntry): boolean => {
```

File: src/storage/calendar.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    if (!calendar.some(elem => entryEq(elem, entry))) {
+      const newCalendar = [...calendar, entry];
+      AsyncStorage.setItem(calendarKey, JSON.stringify(newCalendar))
+        .then(() => setCalendar(newCalendar));
+      return true;
+    }
+    return false;
+  }
+
+  const modifyCalendar = {
```
Comment (Dolvur) [inline src/storage/calendar.tsx]:
```suggestion
  const modifyCalendar: ModifyCalendar = {
```

File: src/storage/icons.tsx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      newIcons[a] = newIcons[b];
+      newIcons[b] = temp;
+      AsyncStorage.setItem(iconsKey, JSON.stringify(newIcons))
+        .then(() => setIcons(newIcons));
+      return true;
+    }
+    return false;
+  }
+
+  const modifyIcons = {
```
Comment (Dolvur) [inline src/storage/icons.tsx]:
```suggestion
  const modifyIcons: ModifyIcons = {
```

File: src/storage/calendar.tsx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+  /**
+   * Adds an entry to the calendar and updates AsyncStorage.
+   *
+   * @param entry - The entry to be added
+   * @returns `true` if the entry was added, `false` otherwise
+   */
+  const add = (entry: CalendarEntry): boolean => {
+    if (!calendar.some(elem => entryEq(elem, entry))) {
+      const newCalendar = [...calendar, entry];
```
Comment (Dolvur) [inline src/storage/calendar.tsx:77]:
```suggestion
      const newCalendar: Calendar = [...calendar, entry];
```

File: src/storage/icons.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+  /**
+   * Adds an icon to the list of icons and updates AsyncStorage.
+   *
+   * @param icon - The icon to be added to the list
+   * @returns `true` if the icon was added, `false` otherwise
+   */
+  const add = (p: string): boolean => {
+    if (!icons.includes(p)) {
+      const newIcons = [...icons, p];
```
Comment (Dolvur) [inline src/storage/icons.tsx]:
```suggestion
      const newIcons: Icons = [...icons, p];
```

File: src/storage/icons.tsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+   * @remark
+   * Swapping the same index will return `false`.
+   *
+   * @param a - The index of the first icon
+   * @param b - The index of the second icon
+   * @returns `true` if the icons were swapped, `false` otherwise
+   */
+  const swap = (a: number, b: number): boolean => {
+    if (a !== b && 0 <= a && a < icons.length && 0 <= b && b < icons.length) {
+      let newIcons = [...icons];
```
Comment (Dolvur) [inline src/storage/icons.tsx:64]:
```suggestion
      let newIcons: Icons = [...icons];
```

File: src/storage/people.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+  /**
+   * Adds a person to the list of people and updates AsyncStorage.
+   *
+   * @param person - The person to be added to the list
+   * @returns `true` if the person was added, `false` otherwise
+   */
+  const add = (person: string): boolean => {
+    if (!people.includes(person)) {
+      const newPeople = [...people, person];
```
Comment (Dolvur) [inline src/storage/people.tsx:40]:
```suggestion
      const newPeople: People = [...people, person];
```

File: src/storage/people.tsx [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+      AsyncStorage.setItem(peopleKey, JSON.stringify(newPeople))
+        .then(() => setPeople(newPeople));
+      return true;
+    }
+    return false;
+  }
+
+  const modifyPeople = {
```
Comment (Dolvur) [inline src/storage/people.tsx]:
```suggestion
  const modifyPeople: ModifyPeople = {
```

File: src/storage/settings.tsx [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+export const useSettings = (): [Settings, ModifySettings] => {
+  const [settings, setSettings] = useState<Settings>(settingsDefault);
+
+  /**
+   * Sets the notifications and updates AsyncStorage.
+   *
+   * @param value - The value to set notifications to
+   */
+  const setNotifications = (value: boolean): void => {
+    let newSettings = { ...settings, notifications: value };
```
Comment (Dolvur) [inline src/storage/settings.tsx:45]:
```suggestion
    let newSettings: Settings = { ...settings, notifications: value };
```

File: src/storage/settings.tsx [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+      .then(() => setSettings(newSettings));
+  }
+
+  /**
+   * Sets the language and updates AsyncStorage.
+   *
+   * @param value - The two letter language code
+   */
+  const setLanguage = (value: string): void => {
+    let newSettings = { ...settings, language: value };
```
Comment (Dolvur) [inline src/storage/settings.tsx:56]:
```suggestion
    let newSettings: Settings = { ...settings, language: value };
```

File: src/storage/settings.tsx [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    AsyncStorage.setItem(settingsKey, JSON.stringify(newSettings))
+      .then(() => setSettings(newSettings));
+  }
+
+  const modifySettings = {
```
Comment (Dolvur) [inline src/storage/settings.tsx]:
```suggestion
  const modifySettings: ModifySettings = {
```

File: src/storage/values.tsx [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+/**
+ * Compares two value entries and returns `true` if they are equal.
+ *
+ * @param a - The first value entry
+ * @param b - The second value entry
+ * @returns a === b
+ */
+const entryEq = (a: ValuesEntry, b: ValuesEntry) => {
```
Comment (Dolvur) [inline src/storage/values.tsx]:
```suggestion
const entryEq = (a: ValuesEntry, b: ValuesEntry): boolean => {
```

File: src/storage/values.tsx [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+   * Adds a topic to the values object and updates AsyncStorage.
+   *
+   * @param category - The category where the topic should be added
+   * @param topic - The name of the topic
+   * @returns `true` if the topic was added, `false` otherwise
+   */
+  const addTopic = (category: string, topic: string): boolean => {
+    if (values.hasOwnProperty(category)) {
+      if (!values[category].some(t => t.name === topic)) {
+        const newValues = JSON.parse(JSON.stringify(values));
```
Comment (Dolvur) [inline src/storage/values.tsx:80]:
```suggestion
        const newValues: Values = JSON.parse(JSON.stringify(values));
```

File: src/storage/values.tsx [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+  /**
+   * Adds an entry to the values object and updates AsyncStorage.
+   *
+   * @param category - The category where the topic should be added
+   * @param topic - The topic in which the entry should be added
+   * @returns `true` if the entry was added, `false` otherwise
+   */
+  const addEntry = (category: string, topic: string, entry: ValuesEntry): boolean => {
+    if (values.hasOwnProperty(category)) {
+      const index = values[category].findIndex(t => t.name === topic);
```
Comment (Dolvur) [inline src/storage/values.tsx:99]:
```suggestion
      const index: number = values[category].findIndex(t => t.name === topic);
```

File: src/storage/values.tsx [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+      if (index !== -1 && !values[category][index].entries.some(e => entryEq(entry, e))) {
+        const newValues = JSON.parse(JSON.stringify(values));
```
Comment (Dolvur) [inline src/storage/values.tsx:101]:
```suggestion
        const newValues: Values = JSON.parse(JSON.stringify(values));
```

File: src/storage/values.tsx [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+        newValues[category][index].entries.push(entry);
+        AsyncStorage.setItem(valuesKey, JSON.stringify(newValues))
+          .then(() => setValues(newValues));
+        return true;
+      }
+    }
+    return false;
+  }
+
+  const modifyValues = {
```
Comment (Dolvur) [inline src/storage/values.tsx]:
```suggestion
  const modifyValues: ModifyValues = {
```

**Review by Dolvur — APPROVED** [review]
Looking good! I added some optional suggestions which you could add if you agree with them. Tested a little with storageTest and addValuesEntry to see why it doesn't work with one click, but it seems like it only renders once (when having multiple setState in one function) if I understand it correctly and therefor the state doesn't get updated for addEntry so the topic to add the entry to doesn't exist

File: src/storage/activities.tsx (hunk 3 above)
Comment (malin-mallan) [reply src/storage/activities.tsx:64]: Agree with Dolvur!

File: src/storage/activities.tsx [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+    return newActivities;
+  }
+
+  /**
+   * Inserts an entry into the activities object and updates AsyncStorage.
+   *
+   * @remarks
+   * This function will add the date if not present.
+   *
+   * @param date - The date on which the entry will be inserted
```
Comment (malin-mallan) [inline src/storage/activities.tsx]: Maybe despcibe how the date should be written? Eg "(YYYY-MM-DD)"

File: src/storage/activities.tsx (hunk 2 above)
Comment (malin-mallan) [reply src/storage/activities.tsx:91]: Agree with Dolvur

File: src/storage/activities.tsx [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  const modifyActivities = {
+    add: add,
+  };
+
+  useEffect(() => {
+    AsyncStorage.getItem(activitiesKey)
+    .then(v => v === null ? activitiesDefault : JSON.parse(v))
+    .then(v => setActivities(v));
```
Comment (malin-mallan) [inline src/storage/activities.tsx]:
```suggestion
    .then(value => value === null ? activitiesDefault : JSON.parse(value))
    .then(value => setActivities(value));
```

File: src/storage/calendar.tsx [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+ * @returns a === b
+ */
+const entryEq = (a: CalendarEntry, b: CalendarEntry) => {
+  return a.date === b.date
+      && a.start === b.start
+      && a.end === b.end
+      && a.text === b.text
+      && a.icon === b.icon
+      && a.person === b.person;
+}
```
Comment (malin-mallan) [inline src/storage/calendar.tsx]: a and b should not have to have an explanation. Use eg calenderEntry1 and calenderEntry2

File: src/storage/calendar.tsx [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  const modifyCalendar = {
+    add: add,
+  };
+
+  useEffect(() => {
+    AsyncStorage.getItem(calendarKey)
+    .then(v => v === null ? calendarDefault : JSON.parse(v))
+    .then(v => setCalendar(v));
```
Comment (malin-mallan) [inline src/storage/calendar.tsx]:
```suggestion
    .then(value => value === null ? calendarDefault : JSON.parse(value))
    .then(value => setCalendar(value));
```

File: src/storage/icons.tsx [hunk 27]
```diff
@@ -0,0 +1,87 @@
+import React, { useState, useEffect } from 'react';
+import AsyncStorage from '@react-native-async-storage/async-storage';
+
+type Icons = string[];
+type ModifyIcons = {
+  add: (p: string) => boolean;
+  swap: (a: number, b: number) => boolean;
```
Comment (malin-mallan) [inline src/storage/icons.tsx]: p, a and b can have better name that describes more

File: src/storage/icons.tsx [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+ * ```
+ * const [icons, modifyIcons] = Storage.useIcons();
+ * modifyIcons.swap(0, 1);
+ * ```
+ */
+export const useIcons = (): [Icons, ModifyIcons] => {
+  const [icons, setIcons] = useState<Icons>([]);
+
+  /**
+   * Adds an icon to the list of icons and updates AsyncStorage.
```
Comment (malin-mallan) [inline src/storage/icons.tsx:37]: Is there a specific icon form required? Eg. JPG? 
If so, it maybe should be described here?

File: src/storage/icons.tsx [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  /**
+   * Swaps two icons in the list and updates AsyncStorage.
+   *
+   * @remark
+   * Swapping the same index will return `false`.
+   *
+   * @param a - The index of the first icon
+   * @param b - The index of the second icon
```
Comment (malin-mallan) [inline src/storage/icons.tsx]: a and b can have better name that describes more

File: src/storage/icons.tsx [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+
+  const modifyIcons = {
+    add: add,
+    swap: swap,
+  };
+
+  useEffect(() => {
+    AsyncStorage.getItem(iconsKey)
+    .then(v => v === null ? iconsDefault : JSON.parse(v))
+    .then(v => setIcons(v));
```
Comment (malin-mallan) [inline src/storage/icons.tsx]:
```suggestion
    .then(value => value === null ? iconsDefault : JSON.parse(value))
    .then(value => setIcons(value));
```

File: src/storage/people.tsx [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+    add: add,
+  };
+
+  useEffect(() => {
+    AsyncStorage.getItem(peopleKey)
+    .then(v => v === null ? peopleDefault : JSON.parse(v))
+    .then(v => setPeople(v));
```
Comment (malin-mallan) [inline src/storage/people.tsx]:
```suggestion
    .then(value => value === null ? peopleDefault : JSON.parse(value))
    .then(value => setPeople(value));
```

File: src/storage/settings.tsx [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+    setNotifications: setNotifications,
+    setLanguage: setLanguage,
+  };
+
+  useEffect(() => {
+    AsyncStorage.getItem(settingsKey)
+    .then(v => v === null ? settingsDefault : JSON.parse(v))
+    .then(v => setSettings(v));
```
Comment (malin-mallan) [inline src/storage/settings.tsx]: v should be value

File: src/storage/values.tsx [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+  health: [],
+  work: [],
+};
+
+/**
+ * Compares two value entries and returns `true` if they are equal.
+ *
+ * @param a - The first value entry
+ * @param b - The second value entry
+ * @returns a === b
```
Comment (malin-mallan) [inline src/storage/values.tsx:39]: a and b should not have to have an explanation. Use eg entry1 and entry2

File: src/storage/values.tsx [hunk 34]
```diff
@@ -0,0 +1,123 @@
...
+
+  const modifyValues = {
+    addTopic: addTopic,
+    addEntry: addEntry,
+  };
+
+  useEffect(() => {
+    AsyncStorage.getItem(valuesKey)
+    .then(v => v === null ? valuesDefault : JSON.parse(v))
+    .then(v => setValues(v));
```
Comment (malin-mallan) [inline src/storage/values.tsx]: Same as above

**Review by malin-mallan — APPROVED** [review]
Really good! 

Use use one letter words on variables a lot which makes the code hard to read. Eg v --> value and a --> entry1 and b --> entry2.

File: src/storage/icons.tsx [hunk 35]
```diff
@@ -3,8 +3,8 @@ import AsyncStorage from '@react-native-async-storage/async-storage';
 
 type Icons = string[];
 type ModifyIcons = {
-  add: (p: string) => boolean;
-  swap: (a: number, b: number) => boolean;
+  add: (icon: string) => boolean;
+  swap: (i1: number, i2: number) => boolean;
```
Comment (malin-mallan) [inline src/storage/icons.tsx:7]: Much better!

**Review by malin-mallan — COMMENTED** [review]
Much better but still a couple of varibles that chould have better namn

**Pull request merged**